_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,234 @@
+
+[hyperband-paper]:https://arxiv.org/pdf/1603.06560.pdf
+[470]:https://scholar.google.com/scholar?cites=10473284631669296057&as_sdt=5,39&sciodt=0,39&hl=en
+
+[^automl]:Their implementation of Hyperband in [HpBandSter] is included in [Auto-PyTorch] and [BOAH].
+
+[^openai]:For OpenAI, NLP preprocessing took 8 GPU-months! ([source][gpt])
+
+[gpt]:https://openai.com/blog/language-unsupervised/#drawbacks
+[Avoiding repeated work]:https://ml.dask.org/hyper-parameter-search.html#avoid-repeated-work
+[BOHB]:https://automl.github.io/HpBandSter/build/html/optimizers/bohb.html
+[hscv]:https://ml.dask.org/modules/generated/dask_ml.model_selection.HyperbandSearchCV.html#dask_ml.model_selection.HyperbandSearchCV
+[BOAH]:https://github.com/automl/BOAH
+[Auto-PyTorch]:https://www.automl.org/wp-content/uploads/2018/09/chapter7-autonet.pdf
+[HpBandSter]:https://github.com/automl/HpBandSter
+[Optuna]:https://medium.com/optuna/optuna-supports-hyperband-93b0cae1a137
+[keras-tune]:https://keras-team.github.io/keras-tuner/documentation/tuners/#hyperband-class
+
+> 3\. **Framework support**: tune-sklearn is used primarily for tuning
+> Scikit-Learn models, but it also supports and provides examples for many
+> other frameworks with Scikit-Learn wrappers such as Skorch (Pytorch),
+> KerasClassifiers (Keras), and XGBoostClassifiers (XGBoost).
+
+This is true for Dask-ML too. However, both Dask-ML and Ray have some
+qualifications. Certain libraries don't offer an implementation of
+`partial_fit`,[^ray-pf] so not all of the modern hyperparameter optimization
+techniques can be offered. That means Bayesian sampling will have to be relied
+upon, which gets less and less relevant as the number of workers grows. Here's
+a table comparing different libraries and their support in Dask-ML and Ray:
+
+| Library | Dask-ML support | Ray support | Dask-ML: early stopping? | Ray: early stopping? |
+|:-----:|:-----:|:-----:|:-----:|:-----:|
+| Scikit-learn | ✔ | ✔ | ✔, with `partial_fit` |✔, with `partial_fit` |
+| PyTorch via Skorch | ✔ | ✔ | ✔ |✔ |
+| Keras | ✔ |✔ | ❌|❌ |
+| LightGDM | ✔ | ✔ | ❌ |❌ |
+| XGBoost |✔ | ✔ | ❌ |❌ |
+
+Of course, Dask has further integration with LightGDM and XGBoost through
+[dask_ml.xgboost][dmlxg] and [dask-lightgdm][dml-lg].
+
+[^ray-pf]:From [Ray's README.md]: "If the estimator does not support `partial_fit`, a warning will be shown saying early stopping cannot be done and it will simply run the cross-validation on Ray's parallel back-end."
+
+[dml-lg]:https://github.com/dask/dask-lightgbm
+[dmlxg]:https://ml.dask.org/xgboost.html
+[Ray's README.md]:https://github.com/ray-project/tune-sklearn/blob/31f228e21ef632a89a74947252d8ad5323cbd043/README.md
+
+> 4\. **Scale up**: Tune-sklearn leverages Ray Tune, a library for distributed
+> hyperparameter tuning, to efficiently and transparently parallelize cross
+> validation on multiple cores and even multiple machines.
+
+Naturally, this is true for Dask-ML too, especially with Dask's wide support of
```

Comment:
Clarify "this" here (that it means "scales to multiple cores, machines") rather than "leverages Ray Tune")